Add flags for TLS Cipher suites customization for API Server, Kubelet and Controller-Manager

rochacon opened the pull request:
This PR adds the TLS Cipher Suite and TLS Min Version flags to core components that support it: apiserver, kubelet and controller-manager.

Fixes #5715

chrisz100 commented:
/ok-to-test

FrederikNS commented:
@rochacon It seems that your commit unfortunately breaks some tests. I'd love to see these changes merged!

chrisz100 commented:
Can you please run `make apimachinery` and commit the changes made by this? That'll fix the issues

rochacon commented:
@chrisz100 thanks. I did a rebase with latest master and ran the make task. I've put the changes in a separate commit for easier review.

Should I add the same field mapping for `pkgs/apis/kops/v1alpha1` and `pkgs/apis/kops/v1alpha2`? I noticed the warnings on the generated code but I don't know the rules on adding new features to those versions.

justinsb commented on upup/pkg/fi/cloudup/openstacktasks/lb.go:
```diff
@@ -18,9 +18,10 @@ package openstacktasks
 	"time"
 
+	"github.com/gophercloud/gophercloud/openstack/networking/v2/ports"
+
```
We probably shouldn't do this in this PR (because it's not thematically related), but if you want to please delete line 24 so that it joins up with the other gophercloud imports.

rochacon replied:
Done, my installed goimports was messing this up for some reason, I removed it from the patch.

justinsb commented on pkg/model/openstackmodel/firewall.go:
```diff
@@ -21,6 +21,7 @@ import (
 	"k8s.io/kops/pkg/apis/kops"
 	"k8s.io/kops/upup/pkg/fi"
 	"k8s.io/kops/upup/pkg/fi/cloudup/openstacktasks"
+
```
We probably shouldn't do this in this PR (because it's not thematically related).  (but not a big deal)

rochacon replied:
Done, my installed goimports was messing this up for some reason, I removed it from the patch. (same as above)

justinsb commented on pkg/apis/kops/componentconfig.go:
```diff
@@ -34,6 +34,10 @@ type KubeletConfigSpec struct {
 	TLSCertFile string `json:"tlsCertFile,omitempty" flag:"tls-cert-file"`
 	// TODO: Remove unused TLSPrivateKeyFile
 	TLSPrivateKeyFile string `json:"tlsPrivateKeyFile,omitempty" flag:"tls-private-key-file"`
+	// TLSCipherSuites indicates the allowed TLS cipher suite
```
So you need to add these fields to v1alpha1 and v1alpha2 (sorry, but that's the way the k8s apimachinery works).  If you search for TLSPrivateKeyFile you'll see the duplication.  Then `make apimachinery` will work :-)

Context: it's possible to change fields in different api versions, and then we would have to write custom conversion functions.  As we're adding the same field to all versions (which is safe) the autogenerated functions work.

rochacon replied:
Done, the new fields are now present on `pkg/apis/kops/component.go`, `pkg/apis/kops/v1alpha1/componentconfig.go` and `pkg/apis/kops/v1alpha2/componentconfig.go`. Code generated by `make apimachinery` seems to be correct now.

justinsb commented on pkg/apis/kops/componentconfig.go:
```diff
@@ -452,6 +460,10 @@ type KubeControllerManagerConfig struct {
 	// FeatureGates is set of key=value pairs that describe feature gates for alpha/experimental features.
 	FeatureGates map[string]string `json:"featureGates,omitempty" flag:"feature-gates"`
+	// TLSCipherSuites indicates the allowed TLS cipher suite
+	TLSCipherSuites []string `json:"tlsCipherSuites,omitempty" flag:"tls-cipher-suites"`
```
Double checked & you're right - the flags are there on kube-controller-manager :+1:

rochacon replied:
I checked the flags on all core components in the 1.11 Kubernetes release binaries. Flag is missing on the kube-scheduler up until Kubernetes 1.13, so I'm ommiting it for now. Should we open an issue to add the flags to it on Kops 1.13?

justinsb replied:
It's fine to map the flags even if they aren't supported by all versions - this is the "advanced" mode, and we can't guarantee compatibility when people start setting flags directly.  If you want to send a PR to add the flags to kube-scheduler, that's great, or an issue if you'd prefer :-)

justinsb commented:
Thanks for this @rochacon - looks really good.  Hopefully the suggestions make sense and get `make apimachinery` working for you!

rochacon commented:
@justinsb thank you for the review, suggestions made complete sense and I've implemented them. I also did a rebase with latest master and squashed the commits (since the output of `make apimachinery` seems correct now).

If build pass, I believe this is good to go.

justinsb commented:
Thanks for fixing that up @rochacon  - looks great

/approve
/lgtm
